### PR TITLE
Add realTime option to http Response

### LINKED ISSUE #4817: SSE support - not doable?
I see in the old deno repo there was a PR in which @ry suggested SSE could instead be implemented outside of std:

https://github.com/denoland/deno_std/pull/91

However I don't think that is doable today because there is no way to flush (short of not using `respond()`/`writeResponse()` but writing the response out to `req.w` manually)

 Is there another way I'm missing?

### PR DESCRIPTION
**Pull request by jakajancar**
Resolves #4817.

### CODE REVIEW

**Comment by hayd** [comment]
Is realTime a standard name? Is that used elsewhere / is there prior art here?

**Comment by jakajancar** [comment]
No, that's my naming. I wanted it to be declarative / descriptive of the response, not imperative / instructive of how `std/http` should handle it. But yeah, I'm not too happy about it either.

.Net has [HttpResponse.BufferOutput](https://docs.microsoft.com/en-us/dotnet/api/system.web.httpresponse.bufferoutput?view=netframework-4.8).

Dart has [HttpResponse.bufferOutput](https://api.dart.dev/stable/2.7.2/dart-io/HttpResponse/bufferOutput.html).

Probably that's more obvious.

**Comment by alisabzevari** [comment]
How about calling it `autoflush`?

File: std/http/io.ts [hunk 1]
```diff
@@ -171,6 +172,9 @@ export async function writeChunkedBody(
     await writer.write(start);
     await writer.write(chunk);
     await writer.write(end);
+    if (options && options.realTime) {
+      await writer.flush();
+    }
```
Comment (alisabzevari) [inline std/http/io.ts:177]: Wouldn't it be possible to make flushing functionality available to callers of `writeChunkedBody`? One example would be sending special values through `r` as a command to flush. To make it more elegant, I would say `response.body` should be a `AsyncIterable`.

**Comment by jakajancar** [comment]
Happy to change when someone with the authority to merge actually makes a decision what is desired (if anything).

**Comment by bartlomieju** [comment]
@jakajancar sorry for slow review; I'll revisit after 1.0 is released

**Comment by jakajancar** [comment]
@bartlomieju Thoughts? I think the `Response` needs a way to indicate it should not be buffered.

**Comment by kitsonk** [comment]
Personal opinion, this is unnecessary.  I commented in #4817 that it is possible with `std/http` today to accomplish SSE without any further capabilities in `std/http`.

**Comment by jakajancar** [comment]
Of course SSE is possible without further capabilities in `std/http` by writing to `.w` directly. The same argument could be made for `std/http` itself: HTTP is possible without further capabilities in `std`, you can write to `Conn` directly :)

Your approach makes sense if you're using the pattern popularized in Node of `handle(req, res)` and functions mutate the request & response.

But if you structure you application more functionally, immutably as:

    handle(req: ServerRequest): Promise<Response>

and handlers never touch the request, only something at the end centrally does `req.respond(res)`, then without having a way to indicate the time-sensitivity of the response in `Response` you have to reimplement response writing functionality that `std/http` otherwise already offers, because it always does buffering.

**Comment by jakajancar** [comment]
Maybe I should rephrase the proposal:

Since `Response` already perfectly supports streaming responses (using a `Reader` as the body), it would be nice if `req.respond()` didn't take away a subset of use-cases of streaming (real-time, not just large data) by always doing buffering for indefinite amounts of time.

To illustrate, this is how I construct my streaming Responses:

```
export interface SSEEvent {
    event?: string
    data?: string
    id?: string
    retry?: number
}

export function sseResponse(events: AsyncIterator<SSEEvent>) {
    return {
        status: Status.OK,
        headers: new Headers({
            'Cache-Control': 'no-cache',
            'Content-Type': 'text/event-stream',
            'X-Accel-Buffering': 'no', // for nginx
        }),
        realTime: true,
        body: toReader(mapAsyncIterator(muxAsyncIterators([events, keepalives(1000)]), serializeEvent))
    }
}
```

**Comment by ry** [comment]
Unfortunately our server API has diverged from the Go http API significantly. These problems are solved in Go well. Rather than continue to diverge, we need to make some large refactors to get back to an API similar to Go's. In particular, the request handler should receive a response writer object (like [`http.ResponseWriter`](https://golang.org/pkg/net/http/#ResponseWriter)) which ensures that when you call `write()` on it, it writes to the underling connection. So unfortunately there's no real easy solution right now. 

Very much appreciate the patch @jakajancar - it's a good solution on top of what we currently have - it's just I think we need deeper changes to fix this situation. I don't think the HTTP server should not be using BufWriter internally.

**Comment by jakajancar** [comment]
Thanks @ry, understandable.

**Pull request closed without merging**